test/jdk/java/net/httpclient/websocket/HandshakeUrlEncodingTest.java

```diff
@@ -0,0 +1,213 @@
+import java.net.URI;
+import java.net.http.HttpClient;
+import java.net.http.WebSocket;
+import java.net.http.WebSocketHandshakeException;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+import static java.net.http.HttpClient.Builder.NO_PROXY;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertNotNull;
+import static org.testng.Assert.fail;
+import static java.lang.System.out;
+
+public class HandshakeUrlEncodingTest {
+
+    SSLContext sslContext;
+    HttpServer httpTestServer;
+    HttpsServer httpsTestServer;
+    String httpURI;
+    String httpsURI;
+
+    static String queryPart;
+
+    static final int ITERATION_COUNT = 10;
+    // a shared executor helps reduce the amount of threads created by the test
+    static final ExecutorService executor = Executors.newCachedThreadPool();
+
+    @DataProvider(name = "variants")
+    public Object[][] variants() {
+        return new Object[][]{
+            { httpURI,   false },
+            { httpsURI,  false },
+            { httpURI,   true  },
+            { httpsURI,  true  }
+        };
+    }
+
+    HttpClient newHttpClient() {
+        return HttpClient.newBuilder()
+                         .proxy(NO_PROXY)
+                         .executor(executor)
+                         .sslContext(sslContext)
+                         .build();
+    }
+
+    @Test(dataProvider = "variants")
+    public void test(String uri, boolean sameClient) {
+        HttpClient client = null;
+        out.println("The url is " + uri);
+        for (int i = 0; i < ITERATION_COUNT; i++) {
+            System.out.printf("iteration %s%n", i);
+            if (!sameClient || client == null)
+                client = newHttpClient();
+
+            try {
+                client.newWebSocketBuilder()
+                    .buildAsync(URI.create(uri), new WebSocket.Listener() { })
+                    .join();
+                fail("Expected to throw");
+            } catch (CompletionException ce) {
+                final Throwable t = getCompletionCause(ce);
+                if (!(t instanceof WebSocketHandshakeException)) {
+                    throw new AssertionError("Unexpected exception", t);
+                }
+                final WebSocketHandshakeException wse = (WebSocketHandshakeException) t;
+                assertNotNull(wse.getResponse());
+                assertNotNull(wse.getResponse().uri());
+                assertNotNull(wse.getResponse().statusCode());
+                final String rawQuery = wse.getResponse().uri().getRawQuery();
+                final String expectedRawQuery = "&raw=abc+def/ghi=xyz&encoded=abc%2Bdef%2Fghi%3Dxyz";
+                assertEquals(rawQuery, expectedRawQuery);
+                // [JDK-8240666] Websocket client's OpeningHandshake discards the HTTP response body
```

Comment:
A better comment would be: `JDK-8240666 is not currently in 11u unlike later JDKs and we, therefore, need to check whether or not a body is present. This is OK as previous assertions verify a fix of JDK-8245245`. Something like that.

Reply:
Ack.